### ISSUE TITLE
Audio Mute not working on tv.apple.com

### ISSUE THREAD

**Issue by joerobbi** [richardfrost/AdvancedProfanityFilter#323]
_Instructions: This template is written in Markdown. Click the "Preview" tab to see what it will look like when submitted. Fill in the applicable sections (you can erase the examples if you'd like)._

#### :bug: Description
Love the idea of this extension. Thanks for creating it! Unfortunately, it doesn't seem to work on Apple TV. I got it specifically to watch Ted Lasso, and it doesn't appear to work. (And yes, I'm using Browser Tab mute method as suggested in a previously posted issue.) F-bombs at 2:38 and 3:36 of the pilot episode land every time.

#### :twisted_rightwards_arrows: Steps To Reproduce
1. Go to tv.apple.com
2. Click on Ted Lasso - Pilot episode
3. Fast Forward to 2:38 (or 3:36)

#### :heavy_check_mark: Expected behavior
As I understand it, the audio should be muted at those times.

#### :camera: Screenshots
![APF_Configuration](https://user-images.githubusercontent.com/87844487/126737712-203d2891-0b8f-4454-9fe3-8bbb8e2e7da1.JPG)

#### :computer: Details
 - **Device:** Windows Desktop
 - **OS:** Windows 10
 - **Browser:** Chrome
 - **Browser Version:** 92.0.4515.107
 - **Affected site(s):** tv.apple.com

**Comment by richardfrost** [comment]
Thanks, I'm glad you like it!

Its possible they could have made an update that broke the filter. I will take a look at it probably this weekend.

Here are some common troubleshooting steps that might be helpful:

- I usually recommend switching the Captions/Subtitles option to "Show all", just to get a better idea of what text is actually there to filter. (After switching that option, you'll need to refresh your Apple TV tab to get the new settings).
- Did you turn the captions on for the video? I seem to remember Apple TV had usually a few options for each language. You might want to try another one and see if you get better results. (When I tested last about a month ago all of the english ones worked fine for me)
- Are the captions being filtered? (If you don't want to hear it you can turn your sound all the way down).
- Can you try temporarily adding a "normal" (non-profane) word (possibly a word in the first sentence, or at an easy-to-find time, like 1:00), and see if it filters that?

**Comment by joerobbi** [comment]
I'm guessing they broke the filter (unless I'm doing something stupid and
don't realize it), because the captions aren't filtered either, even for
custom words. And I do have it turned on. :-)

-Joe

[image: image.png]

**Comment by richardfrost** [comment]
Alright, thanks for testing that! I'll get to it as soon as I have some time to dig in.

**Comment by richardfrost** [comment]
I had a few minutes to start looking into what changed. It looks like they changed how the captions are added to the page, and unfortunately they made it harder for the filter to get them. Here is the best way to get it working again right now:

Go to [https://tv.apple.com](https://tv.apple.com), click on the extension icon, and change the "Mode" to `Deep`.

This will likely cause the page to load slower, because the filter will be processing more than it normally has to. Unfortunately with where they are "hiding" the captions/subtitles this is the only way for the filter to be aware of them right now. 

One other thing to note: They are (now?) re-using the caption/subtitle text, so the filter will only mute the audio the first time showing each caption. For a straight playthrough that should be fine, but if you start seeking around it could cause it to not mute. A page refresh will give you a clean slate again.

It looks like they also changed the way that different captions (for example, the CC version), so those are no longer supported with the default rule. If you wanted it to work for both, putting this in the custom audio sites (Options page > Audio tab) might do the trick:

```json
{
  "tv.apple.com": [
    {
      "displaySelector": "div.video-container > div > div > div",
      "mode": "elementChild",
      "muteMethod": 0,
      "parentSelector": "div.video-container",
      "preserveWhiteSpace": true,
      "rootNode": true,
      "subtitleSelector": "div > div > div > div > div",
      "tagName": "DIV"
    },
    {
      "displaySelector": "div.video-container > div > div > div",
      "mode": "elementChild",
      "muteMethod": 0,
      "parentSelector": "div.video-container",
      "preserveWhiteSpace": true,
      "rootNode": true,
      "subtitleSelector": "div > div > div > div > span",
      "tagName": "DIV"
    }
  ]
}
```

**Comment by richardfrost** [comment]
Quick little update, I've added the above config by default now as of [2.24.0](https://github.com/richardfrost/AdvancedProfanityFilter/releases/tag/2.24.0), so if you want to use the filter all you should need to do is switch `tv.apple.com` to "Deep" mode for it to work.

This is just a partial fix, and I'll see what I can do to help this work out of the box, and with the best performance possible.

**Comment by acougle** [comment]
Exact same problem with Ted Lasso and AppleTV plus. Any updates on this?

**Comment by richardfrost** [comment]
@acougle If you set tv.apple.com to "Deep" mode, does it work for you? The easiest way to do that is click on the extension icon to bring up the popup window while on AppleTV+, and then change the mode to Deep.

**Comment by acougle** [comment]
Thanks so much. Deep setting solved the problem. 
Allan

**Comment by richardfrost** [comment]
Glad to hear it! I do have plans to see if I can come up with a better way to handle it, but for now hopefully this is enough to get by until then.

**Comment by tbeag** [comment]
I still can't seem to figure it out. I have been trying for some days and can't get the filter to work on AppleTV. Is there anyway to get a tutorial of every step from the beginning or something? I know it's a pain, but it would help me a lot.

**Comment by richardfrost** [comment]
I'm sorry you've had so much trouble with it @tbeag. 

You should just need to make sure audio muting is on:

![image](https://user-images.githubusercontent.com/1507205/176079842-797e46a6-ee8d-4699-b509-a8915639722e.png)

And then enable "Deep" mode while on [tv.apple.com](https://tv.apple.com):

![image](https://user-images.githubusercontent.com/1507205/176079961-f9104aba-14fa-4075-b08a-200f78914a68.png)

Let me know how it goes after following those steps.

**Comment by richardfrost** [comment]
Hey @tbeag, just wanted to check in with you and see if you had a chance to try the change mentioned above?

**Comment by tbeag** [comment]
Sorry, I didn’t try for a little bit but it did work!! I was doing way too much on the filter to try and get it to work. I just reset it and did that and it worked. You rock!

**Comment by richardfrost** [comment]
@tbeag Thanks for getting back to me! Hopefully soon I'll be able to get it working out of the box again.

**Comment by Elle-Pea** [comment]
I'm having this same issue with HBOMax. It was working at first, but now it's not even filtering the subtitles. I'm trying to get it to filter both the titles and the audio. 

I have "Mute audio on supported video sites" selected. My mute method is browser tab. My caption selection is "show only filtered." My mode is Deep. I watched the same movie a few days earlier and muting + filtered captioning were both working. I came back to restart the movie again and now neither the captions or audio is being filtered. It's as if APF isn't even working on the page at all because all the subtitles are showing even though I have the filter on "show only filtered." I restarted the movie under another profile in HBOMax because I know you said above that sometimes if you rewind, the filter stops working. Even refreshing, turning the computer off and on, etc. hasn't made it work again. Is there anything I can do to fix it?

**Comment by richardfrost** [comment]
@Elle-Pea Thanks for reporting the issue, and including so many details. There was a really big update to the HBO Max website that went live a couple days ago, and it has indeed broken the functionality of APF to work properly on it. Issue #422 is dedicated to solving this issue, and will have any future updates for it, so you might want to take a look at that. I'm hoping that we can get it taken care of and get it working again soon.

**Comment by Elle-Pea** [comment]
Ok, I'll take a look at #422.

**Comment by emendoza2** [comment]
What needs to be done to get this fixed? I am open to making a pull request if the issue is within my skill set to resolve.

**Comment by richardfrost** [comment]
Thanks for the offer @emendoza2. Are you referring to Apple TV, or HBO Max?

**Comment by emendoza2** [comment]
I'm referring to Apple TV.

**Comment by richardfrost** [comment]
@emendoza2 Thanks! I should have a build that fixes it here in the next day or two. If you'd be willing to test it out I'd appreciate it. Let me know if you're interested.

**Comment by richardfrost** [comment]
This is included in [Version 2.37.0](https://github.com/richardfrost/AdvancedProfanityFilter/releases/tag/2.37.0). If anyone has a chance to test this out, feel free to reply here to report back how it works for you. Thanks!

**IMPORTANT**: If you were using the "Deep" mode as was recommended above, please switch back to "Normal" mode for Apple TV+. If you really want to keep using Deep mode (not recommended), you could add this to your Custom Sites config:

```json
{
  "tv.apple.com": [
    {
      "displaySelector": "div.video-container > div > div > div",
      "mode": "elementChild",
      "muteMethod": 0,
      "parentSelector": "div.video-container",
      "preserveWhiteSpace": true,
      "rootNode": true,
      "subtitleSelector": "div > div > div > div > div",
      "tagName": "DIV"
    },
    {
      "displaySelector": "div.video-container > div > div > div",
      "mode": "elementChild",
      "muteMethod": 0,
      "parentSelector": "div.video-container",
      "preserveWhiteSpace": true,
      "rootNode": true,
      "subtitleSelector": "div > div > div > div > span",
      "tagName": "DIV"
    }
  ]
}
```

Also, if you have Custom Site config for Apple TV+, you should probably remove it as the config included with the extension is now the recommended config.
